zerver/worker/queue_processors.py

```diff
@@ -566,48 +566,54 @@ def consume(self, event: Mapping[str, Any]) -> None:
+    # CHECK_FREQUENCY_SECONDS, to avoid excessive activity.
+    CHECK_FREQUENCY_SECONDS = 5
 
-    def consume(self, event: Dict[str, Any]) -> None:
-        with self.lock:
-            logging.debug("Received missedmessage_emails event: %s", event)
+    worker_thread: Optional[threading.Thread] = None
 
-            # When we consume an event, check if there are existing pending emails
-            # for that user, and if so use the same scheduled timestamp.
-            user_profile_id: int = event["user_profile_id"]
-            user_profile = get_user_profile_by_id(user_profile_id)
-            batch_duration_seconds = user_profile.email_notifications_batching_period_seconds
-            batch_duration = datetime.timedelta(seconds=batch_duration_seconds)
+    # This condition variable mediates the stopping and queue pieces
+    # of state.
+    cv = threading.Condition()
+    stopping = False
 
-            try:
-                pending_email = ScheduledMessageNotificationEmail.objects.filter(
-                    user_profile_id=user_profile_id
-                )[0]
-                scheduled_timestamp = pending_email.scheduled_timestamp
-            except IndexError:
-                scheduled_timestamp = timezone_now() + batch_duration
+    # The main thread, which handles the RabbitMQ connection and creates
+    # database rows from them.
+    def consume(self, event: Dict[str, Any]) -> None:
+        logging.debug("Processing missedmessage_emails event: %s", event)
+        # When we consume an event, check if there are existing pending emails
+        # for that user, and if so use the same scheduled timestamp.
+        user_profile_id: int = event["user_profile_id"]
+        user_profile = get_user_profile_by_id(user_profile_id)
+        batch_duration_seconds = user_profile.email_notifications_batching_period_seconds
+        batch_duration = datetime.timedelta(seconds=batch_duration_seconds)
 
+        try:
+            pending_email = ScheduledMessageNotificationEmail.objects.filter(
+                user_profile_id=user_profile_id
+            )[0]
+            scheduled_timestamp = pending_email.scheduled_timestamp
+        except IndexError:
+            scheduled_timestamp = timezone_now() + batch_duration
+
+        with self.cv:
+            # We now hold the lock, so either:
+            #  - the worker is in maybe_send_batched_emails, and will
+            #    have to take the lock (and thus block insertions of
+            #    new rows) to decide if it needs a timeout
+            #  - the worker is in the cv.wait with a timeout because
+            #    there were rows already
+            #  - the worker is in the cv.wait without a timeout,
+            #    because there weren't any rows (which we're about to
+            #    change)
+            #
+            # Only in the last case do we need to notify.
+            already_pending = ScheduledMessageNotificationEmail.objects.exists()
```

Comment:
I wonder whether this `already_pending` variable would be better implemented as a shared memory variable of some sort that just keeps track of whether the worker is in a timeout-less sleep or not? My read is that's the fact that we're trying to write logic that's conditional on, and it seems like it might be easier to reason about if we can just look at that state rather than trying to predict it based on the state of the database.

(E.g. it'd be nice to not have to think about whether this threading logic will break if we had some other code that mutated `ScheduledMessageNotificationEmail` objects for whatever reason)

Reply:
Sure.  Saves a database call per insert, too!